### ISSUE TITLE
Looping VGMstream on Non-looped songs

### ISSUE THREAD

**Issue by froggestspirit** [droidmjt/Droidsound#376]
I noticed with the soundtracks I play that use VGMstream, as I have on default to play a song through completely twice, then fade out for 5 seconds, this affects everything in VGMstream. Normally this would'nt be an issue, although it starts to behave differently when compared to other formats. MP3, M4A, etc will all play once as intended, emulated formats (such as Nintendo DS's 2SF) will play though twice only if they have loop data. A few songs may not be intended to loop, and will play once, and end without a fade (as intended). However, songs without loop data on VGMstream will still play through twice, then fade (this is mostly noticable in fanfares and such). I noticed that when converting music to wav with VGMstream, it properly loops looped songs, and not the not-looped ones. I breifly looked in the source for the PS3_xvag format, and I believe this to be caused by possibly not reading the "loop_flag" variable. Regardless if loop_flag is set, droidsound-e seems to read the loopStart and loopEnd (as these default to start of the song, and end).

**Comment by droidmjt** [comment]
All right, so how exactly the VGMStreamPlugin should work then?

**Comment by droidmjt** [comment]
Ah, right, so those without loopdata should play once. I'm not sure if it is possible, let's see.

**Comment by droidmjt** [comment]
https://www.dropbox.com/s/t573c7fa9fbvbnx/droidsounde167_48.apk?dl=1

**Comment by froggestspirit** [comment]
That build seemed like it worked some of the time with .xvag files i
tested, I seeked near the end of the first playthrough, sometimes it would
still loop. Although for the times it did work, the length of the song was
still calculated as if it did loop, and would be silent for that portion.
This did not seem to affect other types of files, like DSP and AAX still
looped the songs without loop data.

**Comment by droidmjt** [comment]
Send me those songs you are using for testing.

**Comment by froggestspirit** [comment]
Here are some samples, the only ones I did not test with Droidsound-E are the dsp files (the one I did test was from paper mario TTYD, but it loops when converted with VGStream on the PC). I tested each of these with VGStream on the PC with default looping options, and none of these generated a wav file that loops (as expected).
[no-loopTests.zip](https://github.com/droidmjt/Droidsound/files/1100436/no-loopTests.zip)

**Comment by droidmjt** [comment]
https://www.dropbox.com/s/7nx4064sil67f6g/droidsounde167_49.apk?dl=1

**Comment by froggestspirit** [comment]
I dont believe this one fixed it, still looping.

**Comment by droidmjt** [comment]
In that case, provide clear instructions how to reproduce this looping phenomenon, otherwise, i'll close this case, I'm not able to reproduce it all.

**Comment by froggestspirit** [comment]
Let me try again with cleared settings and see if it makes a difference

**Comment by froggestspirit** [comment]
It seems to not work when fadeout is enabled. When its disabled, songs with
no loop data dont loop

**Comment by droidmjt** [comment]
https://www.dropbox.com/s/0vuaoj6wbsbeg0d/droidsounde167_50.apk?dl=1
test this and let me know problems

**Comment by froggestspirit** [comment]
This one works correctly for the non-looped songs, but it disables fade
settings for the looped songs, including fadeout, and appending the fadeout
to the end. It appears to only affect vgstream songs

**Comment by droidmjt** [comment]
im quite sure i tested fadeout with looped songs

**Comment by froggestspirit** [comment]
Ill try again when im home from work, maybe its only with append fadeout to
end?

**Comment by droidmjt** [comment]
I just tried with the looped songs and it seems to append fadeout just normally. What creates confusion here is that vgmstreamplugin has its own fadeout feature and it always appends the fadeout while the generic fadeout can do both. I need to think what would be proper logic here.

**Comment by droidmjt** [comment]
https://www.dropbox.com/s/uk3kqmwp9fwscrc/droidsounde167_51.apk?dl=1
here, change some logic a bit, hopefully works in every case now

**Comment by froggestspirit** [comment]
Im still seeing a few bugs with this release, in every vgstream song, if
append fadeout is enabled, it appends the fadeout length as silence, adding
extra silence to both looped and non looped songs, while also not fading
the looped ones

**Comment by droidmjt** [comment]
This is how it works now, if you wish to append fadeout, you edit the VGMStreamPlugin fadeout settings, otherwise you use the Fadeout Settings from Settings -> Audio Fade.

**Comment by froggestspirit** [comment]
And if both are set, it should work for all songs regardless of plugin?

**Comment by froggestspirit** [comment]
I did notice that when using the vgstream fadeout, it applies a fadeout to
the non looped songs as well, it doesnt seem to add the fadeout length to
the end, but rather take the last amount of seconds set of the song so it
fades out before the loop ends

On Jun 28, 2017 6:01 AM, "Alec Lafleur" <froggestspirit@gmail.com> wrote:

And if both are set, it should work for all songs regardless of plugin?

**Comment by droidmjt** [comment]
>>it applies a fadeout to the non looped songs as well, it doesnt seem to add the fadeout length to the >>end, but rather take the last amount of seconds set of the song so it fades out before the loop ends
This is expected behaviour because otherwise if you loop a non-looped song with fadeout, it would start again from the beginning and then fade, that's silly. And I think this is what was requested, to not loop non-looped songs.

Anyways, I will remove the vgmstreamplugin fade option so that it uses values from generic fadeout settings.

**Comment by froggestspirit** [comment]
I thought having the two fadouts was a little confusing, although i dont
think fadeout should apply to non looped songs, with most video game
soundtracks, it seems the standard is two playthroughs, then fades out as
the 3rd playthrough starts. Generally the non looped songs either have a
built in fade out, or end in a way where its not needed

On Jun 28, 2017 7:30 AM, "droidmjt" <notifications@github.com> wrote:

it applies a fadeout to the non looped songs as well, it doesnt seem to add
the fadeout length to the >>end, but rather take the last amount of seconds
set of the song so it fades out before the loop ends
This is expected behaviour because otherwise if you loop a non-looped song
with fadeout, it would start again from the beginning and then fade, that's
silly. And I think this is what was requested, to not loop non-looped songs.

Anyways, I will remove the vgmstreamplugin fade option so that it uses
values from generic fadeout settings.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/droidmjt/Droidsound/issues/376#issuecomment-311633098>,
or mute the thread

**Comment by droidmjt** [comment]
https://www.dropbox.com/s/aappymhlk9cwotb/droidsounde167_52.apk?dl=1
try this

**Comment by froggestspirit** [comment]
This build appears to work as intended with all my tests. I appreciate the
work!

**Comment by droidmjt** [comment]
great :)
